Userland/Libraries/LibWeb/Fetch/Response.cpp

```diff
@@ -0,0 +1,309 @@
+        return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, "Status must be in range 200-599"sv };
+
+    // FIXME: 2. If init["statusText"] does not match the reason-phrase token production, then throw a TypeError.
+
+    // 3. Set response’s response’s status to init["status"].
+    m_response->set_status(init.status);
+
+    // 4. Set response’s response’s status message to init["statusText"].
+    m_response->set_status_message(TRY_OR_RETURN_OOM(window, ByteBuffer::copy(init.status_text.bytes())));
+
+    // 5. If init["headers"] exists, then fill response’s headers with init["headers"].
+    if (init.headers.has_value())
+        m_headers->fill(*init.headers);
+
+    // 6. If body was given, then:
+    if (body.has_value()) {
+        // 1. If response’s status is a null body status, then throw a TypeError.
+        if (Infrastructure::is_null_body_status(m_response->status()))
+            return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, "Response with null body status cannot have a body"sv };
+
+        // 2. Set response’s body to body’s body.
+        m_response->set_body(body->body);
+
+        // 3. If body’s type is non-null and response’s header list does not contain `Content-Type`, then append (`Content-Type`, body’s type) to response’s header list.
+        if (body->type.has_value() && m_response->header_list()->contains("Content-Type"sv.bytes())) {
+            auto header = Infrastructure::Header {
+                .name = TRY_OR_RETURN_OOM(global_object(), ByteBuffer::copy("Content-Type"sv.bytes())),
+                .value = TRY_OR_RETURN_OOM(global_object(), ByteBuffer::copy(body->type->span())),
+            };
+            TRY_OR_RETURN_OOM(global_object(), m_response->header_list()->append(move(header)));
+        }
+    }
+
+    return {};
+}
+
+// https://fetch.spec.whatwg.org/#dom-response
+WebIDL::ExceptionOr<JS::NonnullGCPtr<Response>> Response::create_with_global_object(HTML::Window& window, Optional<BodyInit> const& body, ResponseInit const& init)
+{
+    auto& realm = window.realm();
+
+    // Referred to as 'this' in the spec.
+    auto response_object = [&] {
+        auto response = adopt_own(*new Infrastructure::Response());
+        return JS::NonnullGCPtr { *realm.heap().allocate<Response>(realm, realm, move(response)) };
+    }();
+
+    // 1. Set this’s response to a new response.
+    // NOTE: This is done at the beginning as the 'this' value Response object
+    //       cannot exist with a null Infrastructure::Response.
+
+    // 2. Set this’s headers to a new Headers object with this’s relevant Realm, whose header list is this’s response’s header list and guard is "response".
+    response_object->m_headers = realm.heap().allocate<Headers>(realm, window);
+    response_object->m_headers->set_header_list(response_object->response().header_list());
+    response_object->m_headers->set_guard(Headers::Guard::Response);
```

Comment:
Here not again? (I could be missing something obvious though)

Reply:
Same as above, "list is this’s response’s header" means copying the RefPtr.